### PR TITLE
install: introduce installer class hierarchy

### PR DESCRIPTION
**Pull request by jcholast**

### CODE REVIEW

File: ipapython/install/cli.py [hunk 1]
```diff
@@ -112,7 +112,7 @@ def add_options(cls, parser):
 
         for owner_cls, name in transformed_cls.knobs():
             knob_cls = getattr(owner_cls, name)
-            if knob_cls.cli_positional:
+            if knob_cls.cli_positional is not positional:
```
Comment (mbasti-rh) [inline ipapython/install/cli.py]: It is not clean to me why do you separate positional and non-positional options, and why it is done in this way

Comment (mbasti-rh) [inline ipapython/install/cli.py]: Options are non positional, so probably for better reading, there should be new method `add_arguments`

**Comment by mbasti-rh** [comment]
I reviewed:
- install: improve positional argument handling
- install: simplify CLI option parsing
- install: rename Knob to knob

All of them needs ticket and proper description and reasoning in commit message

File: ipapython/install/cli.py [hunk 2]
```diff
@@ -23,20 +23,31 @@
...
+    for owner_cls, name in configurable_class.knobs():
+        knob_cls = getattr(owner_cls, name)
+        if knob_cls.cli_positional:
+            if knob_cls.cli_metavar is not None:
+                metavar = knob_cls.cli_metavar
+            elif knob_cls.cli_name is not None:
+                metavar = knob_cls.cli_name.upper()
+            else:
+                metavar = name.replace('_', '-').upper()
+            usage += ' [{}]'.format(metavar)
```
Comment (jcholast) [inline ipapython/install/cli.py]: I don't. Fixed.

File: ipapython/install/cli.py (hunk 1 above)
Comment (jcholast) [inline ipapython/install/cli.py]: Added explanatory comments to the relevant parts of the code.

**Comment by mbasti-rh** [comment]
```
[root@vm-058-017 ~]# ipa-server-install -U --uninstall
Traceback (most recent call last):
  File "/sbin/ipa-server-install", line 37, in <module>
    ServerInstall.run_cli()
  File "/usr/lib/python2.7/site-packages/ipapython/admintool.py", line 125, in run_cli
    sys.exit(cls.main(sys.argv))
  File "/usr/lib/python2.7/site-packages/ipapython/admintool.py", line 141, in main
    cls.make_parser()
  File "/usr/lib/python2.7/site-packages/ipapython/admintool.py", line 100, in make_parser
    cls.add_options(parser)
  File "/usr/lib/python2.7/site-packages/ipapython/install/cli.py", line 294, in add_options
    super(InstallTool, cls).add_options(parser, positional)
  File "/usr/lib/python2.7/site-packages/ipapython/install/cli.py", line 175, in add_options
    **kwargs
  File "/usr/lib64/python2.7/optparse.py", line 1013, in add_option
    option = self.option_class(*args, **kwargs)
  File "/usr/lib64/python2.7/optparse.py", line 578, in __init__
    checker(self)
  File "/usr/lib64/python2.7/optparse.py", line 661, in _check_type
    raise OptionError("invalid option type: %r" % self.type, self)
optparse.OptionError: option --ip-address: invalid option type: 'ip-local'

```

**Comment by mbasti-rh** [comment]
Needs rebase

**Comment by mbasti-rh** [comment]
I again reviewved just first three commits as before and I got:

```
ipa-server-install -h
Traceback (most recent call last):
  File "/sbin/ipa-server-install", line 37, in <module>
    ServerInstall.run_cli()
  File "/usr/lib/python2.7/site-packages/ipapython/admintool.py", line 125, in run_cli
    sys.exit(cls.main(sys.argv))
  File "/usr/lib/python2.7/site-packages/ipapython/admintool.py", line 141, in main
    cls.make_parser()
  File "/usr/lib/python2.7/site-packages/ipapython/admintool.py", line 100, in make_parser
    cls.add_options(parser)
  File "/usr/lib/python2.7/site-packages/ipapython/install/cli.py", line 302, in add_options
    super(InstallTool, cls).add_options(parser, positional)
  File "/usr/lib/python2.7/site-packages/ipapython/install/cli.py", line 183, in add_options
    **kwargs
  File "/usr/lib64/python2.7/optparse.py", line 1013, in add_option
    option = self.option_class(*args, **kwargs)
  File "/usr/lib64/python2.7/optparse.py", line 578, in __init__
    checker(self)
  File "/usr/lib64/python2.7/optparse.py", line 661, in _check_type
    raise OptionError("invalid option type: %r" % self.type, self)
optparse.OptionError: option --ip-address: invalid option type: 'ip-local'
```

**Comment by mbasti-rh** [comment]
INFO related to LDAPError:

```
-> desc = e.args[0]['desc'].strip()
(Pdb) print e
LDAPError(2, 'No such file or directory')
(Pdb) print self.ldap_uri
'master.ipa.test'
```
Probaly we should open ticket and improve LDAP error handling to cover this issue

error is in test_integration/host.py:ldap_connect
```
        ldap_uri = ipaldap.LDAPClient(self.external_hostname)
        ldap = ipaldap.LDAPClient(ldap_uri)  <-------- you created ldap_uri as LDAPClient object
```

**Comment by mbasti-rh** [comment]
NACK

```
************* Module ipalib.install.hostname
ipalib/install/hostname.py:5: [E0401(import-error), ] Unable to import 'typing')
************* Module ipalib.install.service
ipalib/install/service.py:5: [E0401(import-error), ] Unable to import 'typing')
************* Module ipaclient.install.client
ipaclient/install/client.py:27: [E0401(import-error), ] Unable to import 'typing')
************* Module ipaserver.install.ca
ipaserver/install/ca.py:9: [E0401(import-error), ] Unable to import 'typing')
************* Module ipaserver.install.dns
ipaserver/install/dns.py:9: [E0401(import-error), ] Unable to import 'typing')
************* Module ipaserver.install.server.common
ipaserver/install/server/common.py:31: [E1101(no-member), BaseServerCA] Module 'ipapython.install.core' has no 'Group' member)
ipaserver/install/server/common.py:162: [E1101(no-member), BaseServerDNS] Module 'ipapython.install.core' has no 'Group' member)
************* Module ipaserver.install.server
ipaserver/install/server/__init__.py:11: [E0401(import-error), ] Unable to import 'typing')
```

**Comment by mbasti-rh** [comment]
PEP8 errors
```
./ipaserver/install/server/__init__.py:493:5: E129 visually indented line with same indent as next logical line
./ipaserver/install/server/__init__.py:493:21: W503 line break before binary operator
./ipaserver/install/server/__init__.py:590:21: W503 line break before binary operator

./ipapython/install/cli.py:157:33: E721 do not compare types, use 'isinstance()'

```

**Comment by mbasti-rh** [comment]
A.k.a as we agreed offline, typing module should not be used

File: ipaserver/install/server/common.py [hunk 3]
```diff
@@ -200,7 +200,9 @@ class BaseServerDNS(common.Installable, core.Group, core.Composite):
 
     @reverse_zones.validator
     def reverse_zones(self, values):
-        if not self.allow_zone_overlap:
+        try:
```
Comment (mbasti-rh) [inline ipaserver/install/server/common.py]: Why this change is needed for Flags?  It doesnt look for me to catch Attribute error instead comparing True/False for flags

**Comment by mbasti-rh** [comment]
```
[root@vm-058-017 ~]# ipa-server-install -a hesloheslo -p hesloheslo --setup-dns

<Fill requested info>

ipa.ipapython.install.cli.install_tool(StandaloneServerMasterInstall): ERROR    dnssec_master
ipa.ipapython.install.cli.install_tool(StandaloneServerMasterInstall): ERROR    The ipa-server-install command failed. See /var/log/ipaserver-install.log for more information

```

Traceback
```
  File "/usr/lib/python2.7/site-packages/ipaserver/install/server/__init__.py", line 462, in main
    master_install_check(self)
  File "/usr/lib/python2.7/site-packages/ipaserver/install/server/install.py", line 264, in decorated
    func(installer)
  File "/usr/lib/python2.7/site-packages/ipaserver/install/server/install.py", line 603, in install_check
    dns.install_check(False, api, False, options, host_name)
  File "/usr/lib/python2.7/site-packages/ipaserver/install/dns.py", line 174, in install_check
    if (options.dnssec_master and not options.unattended and not
  File "/usr/lib/python2.7/site-packages/ipapython/install/core.py", line 662, in __getattr__
    raise AttributeError(name)

```

**Comment by mbasti-rh** [comment]
With these patches, installers are extremely slow, I thinkt hat processing konobs and all the clasess + inheritance, causes slowness of installer, `ipa-client-install -U --uninstall` took about 15 seconds until it start doing something. I can do profiling if you want.

**Comment by mbasti-rh** [comment]
Okay it was a temporal issue, now it work fast again (when I attached profiler). Knobs processing took just miliseconds

**Comment by mbasti-rh** [comment]
```
ipa-replica-install

...Configuring ipa-otpd
  [1/2]: starting ipa-otpd 
  [2/2]: configuring ipa-otpd to start on boot
Done configuring ipa-otpd.
Your system may be partly configured.
Run /usr/sbin/ipa-server-install --uninstall to clean up.

ipa.ipapython.install.cli.install_tool(StandaloneServerSlaveInstall): ERROR    can't set attribute
ipa.ipapython.install.cli.install_tool(StandaloneServerSlaveInstall): ERROR    The ipa-replica-install command failed. See /var/log/ipareplica-install.log for more information



  File "/usr/lib/python2.7/site-packages/ipapython/install/common.py", line 63, in _install
    for _nothing in self._installer(self.parent):
  File "/usr/lib/python2.7/site-packages/ipaserver/install/server/__init__.py", line 553, in promote_main
    promote(self)
  File "/usr/lib/python2.7/site-packages/ipaserver/install/server/replicainstall.py", line 408, in decorated
    func(installer)
  File "/usr/lib/python2.7/site-packages/ipaserver/install/server/replicainstall.py", line 1541, in promote
    options.dm_password = config.dirman_password

2016-11-08T16:29:31Z DEBUG The ipa-replica-install command failed, exception: AttributeError: can't set attribute
2016-11-08T16:29:31Z ERROR can't set attribute
2016-11-08T16:29:31Z ERROR The ipa-replica-install command failed. See /var/log/ipareplica-install.log for more information

```

**Comment by mbasti-rh** [comment]
I don't like changes to Master/Slave in class names, can we stay with master/replica?

**Comment by mbasti-rh** [comment]
I miss descriptions in Installer classes, I have no idea what class has which meaning
why do we have separation to : install, enroll, prepare interfaces? Most of the time I see that prepare and enroll are merged in one empty class later, and then used in another class? Would be possible to reduce amount of inheritance?

**Comment by jcholast** [comment]
@mbasti-rh, fixed the bugs you reported and reduced the amount of inheritance.

Regarding master/replica, it does not make much sense, because master is a replica too. I'm open to other suggestions though.

**Comment by mbasti-rh** [comment]
ACK

**Comment by jcholast** [comment]
Pushed to installer-refactoring.
